File: android/core/domain/src/main/java/com/enterprise/core/domain/usecases/DeepLinkParser.kt
package com.enterprise.core.domain.usecases

import android.net.Uri
import com.enterprise.core.domain.entities.AppRoute
import javax.inject.Inject

/** Parses URI strings into strongly-typed AppRoute navigation intents. */
class DeepLinkParser @Inject constructor() {

    fun parse(uriString: String): AppRoute {
        if (uriString.isBlank()) return AppRoute.Unknown(uriString)

        val uri = try {
            Uri.parse(uriString)
        } catch (_: Exception) {
            return AppRoute.Unknown(uriString)
        }

        val pathSegments = uri.pathSegments ?: emptyList()
        val host = uri.host ?: return AppRoute.Unknown(uriString)

        return when (host) {
            "explore" -> AppRoute.Explore
            "shop" -> parseShopRoute(pathSegments, uri)
            "account" -> AppRoute.Account
            "rewards" -> parseRewardsRoute(pathSegments)
            "chat" -> parseChatRoute(pathSegments)
            "search" -> AppRoute.Search(uri.getQueryParameter("q"))
            "notification" -> {
                val id = pathSegments.firstOrNull() ?: return AppRoute.Unknown(uriString)
                AppRoute.Notification(id)
            }
            else -> AppRoute.Unknown(uriString)
        }
    }

    private fun parseShopRoute(segments: List<String>, uri: Uri): AppRoute {
        if (segments.isEmpty()) return AppRoute.Shop

        return when (segments[0]) {
            "cart" -> {
                if (segments.size > 1 && segments[1] == "add") {
                    val itemId = uri.getQueryParameter("itemId") ?: return AppRoute.CartView
                    AppRoute.CartAdd(itemId)
                } else {
                    AppRoute.CartView
                }
            }
            else -> AppRoute.ProductDetail(segments[0])
        }
    }

    private fun parseRewardsRoute(segments: List<String>): AppRoute {
        return if (segments.isEmpty()) {
            AppRoute.Rewards
        } else {
            AppRoute.RewardDetail(segments[0])
        }
    }

    private fun parseChatRoute(segments: List<String>): AppRoute {
        return AppRoute.ChatSession(segments.firstOrNull())
    }
}

File: ios/EnterpriseMobileApp/Sources/Core/Domain/UseCases/DeepLinkParser.swift
import Foundation

/// Centralized deep link parser translating URIs into domain-level routing intents.
public struct DeepLinkParser: Sendable {
    public init() {}

    public func parse(_ urlString: String) -> AppRoute {
        guard let url = URL(string: urlString),
              let host = url.host else {
            return .unknown(path: urlString)
        }

        let pathComponents = url.pathComponents.filter { $0 != "/" }
        let queryItems = URLComponents(url: url, resolvingAgainstBaseURL: false)?
            .queryItems?.reduce(into: [String: String]()) { dict, item in
                dict[item.name] = item.value
            } ?? [:]

        switch host {
        case "explore":
            return .explore
        case "shop":
            if pathComponents.isEmpty { return .shop }
            switch pathComponents[0] {
            case "cart":
                if pathComponents.count > 1 && pathComponents[1] == "add",
                   let itemId = queryItems["itemId"] {
                    return .cartAdd(itemId: itemId)
                }
                return .cartView
            default:
                return .productDetail(productId: pathComponents[0])
            }
        case "account":
            return .account
        case "rewards":
            if let rewardId = pathComponents.first {
                return .rewardDetail(rewardId: rewardId)
            }
            return .rewards
        case "chat":
            let sessionId = pathComponents.first
            return .chatSession(sessionId: sessionId)
        case "search":
            return .search(query: queryItems["q"])
        case "notification":
            if let notifId = pathComponents.first {
                return .notification(notificationId: notifId)
            }
            return .unknown(path: urlString)
        default:
            return .unknown(path: urlString)
        }
    }
}
